mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAOptimizer.scala

```diff
@@ -208,3 +225,224 @@ class EMLDAOptimizer extends LDAOptimizer{
+  private var Elogbeta: BDM[Double]= null
+  private var expElogbeta: BDM[Double] = null
+
+  // count of invocation to next, used to help deciding the weight for each iteration
+  private var iteration = 0
+
+  /**
+   * A (positive) learning parameter that downweights early iterations
+   */
+  def getTau_0: Double = {
+    if (this.tau_0 == -1) {
+      1024
+    } else {
+      this.tau_0
+    }
+  }
+
+  /**
+   * A (positive) learning parameter that downweights early iterations
+   * Automatic setting of parameter:
+   *  - default = 1024, which follows the recommendation from OnlineLDA paper.
+   */
+  def setTau_0(tau_0: Double): this.type = {
+    require(tau_0 > 0 || tau_0 == -1.0,  s"LDA tau_0 must be positive, but was set to $tau_0")
+    this.tau_0 = tau_0
+    this
+  }
+
+  /**
+   * Learning rate: exponential decay rate
+   */
+  def getKappa: Double = {
+    if (this.kappa == -1) {
+      0.5
+    } else {
+      this.kappa
+    }
+  }
+
+  /**
+   * Learning rate: exponential decay rate---should be between
+   * (0.5, 1.0] to guarantee asymptotic convergence.
+   *  - default = 0.5, which follows the recommendation from OnlineLDA paper.
+   */
+  def setKappa(kappa: Double): this.type = {
+    require(kappa >= 0 || kappa == -1.0,
+      s"Online LDA kappa must be nonnegative (or -1 for auto), but was set to $kappa")
+    this.kappa = kappa
+    this
+  }
+
+  /**
+   * Mini-batch size, which controls how many documents are used in each iteration
+   */
+  def getBatchSize: Int = {
+    if (this.batchSize == -1) {
+      D / 100
+    } else {
+      this.batchSize
+    }
+  }
+
+  /**
+   * Mini-batch size, which controls how many documents are used in each iteration
+   * default = 1% from total documents.
+   */
+  def setBatchSize(batchSize: Int): this.type = {
+    this.batchSize = batchSize
+    this
+  }
+
+  private[clustering] override def initialize(docs: RDD[(Long, Vector)], lda: LDA): LDAOptimizer = {
+
+    this.k = lda.getK
+    this.D = docs.count().toInt
+    this.vocabSize = docs.first()._2.size
+    this.alpha = if (lda.getDocConcentration == -1) 1.0 / k else lda.getDocConcentration
+    this.eta = if (lda.getTopicConcentration == -1) 1.0 / k else lda.getTopicConcentration
+    this.randomSeed = randomSeed
+
+    this.docs = docs
+
+    // Initialize the variational distribution q(beta|lambda)
+    this.lambda = getGammaMatrix(k, vocabSize)
+    this.Elogbeta = dirichlet_expectation(lambda)
+    this.expElogbeta = exp(Elogbeta)
+    this.iteration = 0
+    this
+  }
+
+  /**
+   * Submit a a subset (like 1%, decide by the batchSize) of the corpus to the Online LDA model,
+   * and it will update the topic distribution adaptively for the terms appearing in the subset.
+   *
+   * @return  Inferred LDA model
+   */
+  private[clustering] override def next(): OnlineLDAOptimizer = {
+    iteration += 1
+    val batchSize = getBatchSize
+    val batch = docs.sample(true, batchSize.toDouble / D, randomSeed).cache()
```

Comment:
This batch is only used once, so you won't gain anything by caching it.

randomSeed: Currently, this will use the same sample on every iteration.  You should create a java.util.Random instance, set the seed using randomSeed, and then use it on each iteration to create a new random seed to pass to sample().